sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenRefreshOptions.java

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.credential;
+
+import java.time.Duration;
+
+/**
+ * The options to configure the token refresh behavior.
+ */
+public class TokenRefreshOptions {
+    private static final Duration DEFAULT_REFRESH_RETRY_TIMEOUT = Duration.ofSeconds(30);
+    private static final Duration DEFAULT_REFRESH_OFFSET = Duration.ofMinutes(2);
+
+    private Duration tokenRefreshRetryTimeout = DEFAULT_REFRESH_RETRY_TIMEOUT;
+    private Duration tokenRefreshOffset = DEFAULT_REFRESH_OFFSET;
+
+    /**
+     * Returns a Duration value representing the amount of time to wait before retrying a token refresh. This is to
+     * prevent sending too many requests to the authentication service.
+     *
+     * @return the duration value representing the amount of time to wait before retrying a token refresh
+     */
+    public Duration getTokenRefreshRetryTimeout() {
+        return tokenRefreshRetryTimeout;
+    }
+
+    /**
+     * Specifies a Duration value representing the amount of time to wait before retrying a token refresh. This is to
+     * prevent sending too many requests to the authentication service.
+     *
+     * @param tokenRefreshRetryTimeout the amount of time to wait before retrying a token refresh
+     * @return the updated TokenRefreshOptions object
+     */
+    public TokenRefreshOptions setTokenRefreshRetryTimeout(Duration tokenRefreshRetryTimeout) {
```

Comment:
I think it's could be problematic to have setters for these properties. If object is mutable `TokenCredential` implementations will have to copy this object to avoid mutations in behavior.  Is it possible to only take these arguments on construction? or would we need to add a builder to make this class immutable?

Reply:
I can move them to a constructor but then it will be difficult to add more options in the future. Most of the "XXXOptions" in the Java SDK today are mutable like [AmqpRetryOptions](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/AmqpRetryOptions.java), [HttpLogOptions](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/policy/HttpLogOptions.java)

Reply:
From the Java SDK authoring guidelines (https://azure.github.io/azure-sdk/java_design.html#model-classes):

> ✅ DO provide public constructors for all model classes that a user is allowed to instantiate. Model classes that are not instantiable by the user, for example if they are model types returned from the service, should not have any publicly visible constructors.
> 
> Use a no-args constructor and a fluent setter API to configure the model class. However, other designs may be used for the constructor when appropriate.